---
title: "Regulatory network inference"
author: "Health Data Unit"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: hide
  pdf_document: default
---


# Intro

This report contain documentation of the analysis of CD8 T cells from chronic HCV
patients.


```{r, setting, include=FALSE}
library(rio)
library(ggplot2)
library(tidyverse)
#library(motifmatchr)
#library(TFBSTools)
library(JASPAR2020)
library(GenomicRanges)
#library(BSgenome)
library(ggVennDiagram)
library(grid)
#library(Gviz)
#library(Homo.sapiens)
library(rGREAT)
library(UpSetR)
library(ggrepel)
library(Seurat)
library(reshape2)
library(ggpubr)
library(igraph)
library(ComplexHeatmap)
library(viridis)
library(AUCell)
library(ggrepel)

## Rmarkdown Settings
knitr::opts_chunk$set(
        warning = FALSE, 
        message = FALSE,
        cache = FALSE,
        fig.align = 'center'
)

set.seed(333)

path2project <- '/media/ag-cherrmann/cramirez/tcd8ExscSeq/'
```


# Reading peaks

* The data was taken from [Kathleen B. Y. et al, 2021](https://www.nature.com/articles/s41590-021-00979-1?proof=t#additional-information). 


The next table shows the universe of ChARs identified in all conditions.


```{r, loading_peaks}
peaks_url <- 'https://static-content.springer.com/esm/art%3A10.1038%2Fs41590-021-00979-1/MediaObjects/41590_2021_979_MOESM3_ESM.xlsx'

download.file(peaks_url, '/media/ag-cherrmann/cramirez/tcd8ExscSeq/data/41590_2021_979_MOESM3_ESM.xlsx')
peaks.df <- readxl::read_xlsx('/media/ag-cherrmann/cramirez/tcd8ExscSeq/data/41590_2021_979_MOESM3_ESM.xlsx', 
                              sheet = 'Supplementary Table 2')
#dim(peaks.df)
head(peaks.df)
```


# The epigenetic scar

```{r, size_of_epigenetic_scar}
universe <- pull(peaks.df, ID)
preVspost.HCV <- filter(peaks.df, 
                        as.numeric(`q-value: Group HCV_pre vs HCV_post`) < 0.05) %>% pull(ID)
sets <- list(universe=universe,
             `Pre- Vs Post-Treatment`=preVspost.HCV)
```


The next Venn Diagram shows the number of CHaRs in the overall universe across all conditions 
(`r length(universe)`) and the number of differential peaks in the epigenetic scar which 
corresponds to the comparison of Pre- vs Post-DAA treatment (`r length(preVspost.HCV)`).


```{r}
ggVennDiagram(sets)
```

The next vulcano plot shows the epigenetic scar.

```{r, fig.height=4, fig.width=4}
## Visualization of the epigenetic scar
epi.scar.vulcano.plot <- peaks.df %>%
  ggplot(aes(x=`log2FC: Group HCV_pre vs HCV_post`,
             y=-log10(as.numeric(`q-value: Group HCV_pre vs HCV_post`)))) +
       geom_point(alpha=0.3, size=3) +
       geom_hline(yintercept = -log10(0.05), linetype='dashed', 
                  color='red') +
       theme_classic() +
       labs(x='Log2FC: HCV pre Vs post', y='-Log10(q-Value)') +
       annotation_custom(
         grobTree(textGrob('q-Value=0.05', x=0.2, y=0.1),
                gp=gpar(col='red'))
       )
       
       
epi.scar.vulcano.plot
```

# Scanning motifs in CHaRs

We used MotifMatchR to scan the motifs in open regions. Motifs used
correspond to the JASPAR2020 database.


```{r, scanning_motifs_motif_matchr, eval=FALSE}
## Constructing the set of PFmatrix containing all the motifs in jaspar2020
opts <- list()
opts[["all_versions"]] <- TRUE
PFMatrixList <- getMatrixSet(JASPAR2020, 
                             opts)

## Converting regions to GRanges
peaks <- peaks.df
#peaks <- filter(peaks.df, 
#                as.numeric(`q-value: Group HCV_pre vs HCV_post`) < 0.05)
ranges <- makeGRangesFromDataFrame(peaks, 
                                   ignore.strand = TRUE, 
                                   seqnames.field = 'Chr', 
                                   start.field = 'Start',
                                   end.field = 'End')


motif_ix <- matchMotifs(PFMatrixList, 
                        ranges, 
                        genome = "hg19", 
                        out = 'scores') 

#saveRDS(motif_ix, paste0(path2project,
#                         '/analysis/motifs_CHaRs_preVsPost_treatment.rds'),
#        compress = TRUE)
#saveRDS(ranges, paste0(path2project, 
#                       'analysis/ranges_CHaRs_preVsPost_treatment.rds'),
#        compress = TRUE)
dim(motifScores(motif_ix))
head(motifScores(motif_ix))
```

# Plotting an example of ChAR 

The next track plot show an interval of 100 kb around the IFNG gene.

```{r, char_plot, eval=FALSE}
ranges <- readRDS(file = paste0(path2project,
                         '/analysis/ranges_CHaRs_preVsPost_treatment.rds'))
motif_ix <- readRDS(file = paste0(path2project,
                         '/analysis/motifs_CHaRs_preVsPost_treatment.rds'))
from <- 68548548 - 100000 
to <- 68553520 + 100000
ranges.12 <- ranges[seqnames(ranges)=="chr12" &
                      from < start(ranges)  &
                         end(ranges) < to]
atrack <- AnnotationTrack(ranges.12)
gtrack <- GenomeAxisTrack()
chr <- 'chr12'
gen <- 'hg19'
itrack <- IdeogramTrack(genome = gen, 
                        chromosome = chr)


knownGenes <- UcscTrack(genome = "hg19", 
                        chromosome = "chr12", 
                        track = "knownGene", 
                        from = from, to = to,
                        trackType = "GeneRegionTrack", 
                        rstarts = "exonStarts", 
                        rends = "exonEnds", 
                        gene = "name", 
                        symbol = "name", 
                        transcript = "name", 
                        strand = "strand", 
                        fill = "#8282d2", 
                        name = "UCSC Genes")

plotTracks(list(itrack, atrack, gtrack, knownGenes))
```

# Annotating regions with genes around regions

We annotate regions with the closest genes. We calculated the distances
of the region to all the gRanges regions in the UCSC hg19 database of
known genes. And then we kept just the genes with less than a 1 kb.

We tested different distance windows (1, 3, 5, 10, and 50 kb).
The following upset plot shows the set of genes mapped to the CHaRs in the ATAC-Seq data found using the different distance windows.


```{r, manual_annotation, eval=FALSE}
## Construction of the annotated reference
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene::TxDb.Hsapiens.UCSC.hg19.knownGene
broads <- GenomicFeatures::genes(txdb)
subject <- broads[ seqnames(broads) %in% seqlevels(ranges) ]

## Construction of the gene annotations
TxDb(Homo.sapiens) <- txdb
tx.list <- transcriptsBy(Homo.sapiens, columns = "SYMBOL")
tx <- unlist(tx.list)

## Function to annotate regions to genes around 1, 3, 10, 50 and 100 kbs
getNeighborsGenes <- function(grange, txDb_reference, dist){
  tx.chr <- txDb_reference[seqlevels(txDb_reference) %in% seqlevels(grange)]
  distances <- GenomicRanges::distance(grange, tx.chr)
  names(distances) <- unlist(tx.chr$SYMBOL)
  distances <- distances[!is.na(distances) & ! is.na(names(distances))]
  return(distances[abs(distances)<dist])
}

distances <- c(1000, 3000, 5000, 10000, 50000)
neighbors.list <- lapply(distances, 
       function(distance_val){
         lapply(1:length(ranges), 
                    function(i){ 
                      getNeighborsGenes(ranges[i], 
                                        txDb_reference = tx, 
                                        dist = distance_val)})
})
names(neighbors.list) <- paste0('neighborGenes_', distances)
#lapply(neighbors.list, head)

granges.genes <- lapply(neighbors.list, 
                        function(res){
                          genes <- res %>% unlist() %>% names() %>% unique()
                        })
names(granges.genes) <- paste0('genes_around_', c(1, 3, 5, 10, 50), '_kb')

```


```{r, eval=FALSE}
upset(fromList(granges.genes), nsets = 20)
```


It can be observed that most of the mapped genes are located at less than 1 kb
of distance to the corresponding associated peak.


```{r, distances, eval=FALSE}
distances.filtered <- neighbors.list$neighborGenes_1000 %>% 
                             unlist() 
sum(distances.filtered==0)
hist(distances.filtered[distances.filtered> 0], breaks = 10)
```



```{r, match_motif_annotation, eval=FALSE}
## getting mapped regions
map.genByRegion <- sapply(neighbors.list$neighborGenes_1000,
        function(genes){
          genes.unique <- names(genes) %>% unique() 
          if (length(genes.unique) > 0) {
            res <- paste(genes.unique, collapse = ';')
          } else {
            res <- ''
          }
          return(res)
}) 



## Getting TFs mapped to regions
tfs.bound <- character(length = nrow(motif_ix))
mtx <- motifMatches(motif_ix)
for (i in 1:nrow(motif_ix)){
  tfs.bound[i] <- paste(motif_ix@colData$name[mtx[i,]],
                        collapse = ';')
  cat('Region', i, 'has been processed\n')
}
head(tfs.bound)


## Adding region mapping to found motifs in regions
values(motif_ix) <- DataFrame(genes=map.genByRegion,
                              TF=tfs.bound)

saveRDS(motif_ix, paste0(path2project,
                         '/analysis/motifs_CHaRs_preVsPost_treatment.rds'),
        compress = TRUE)

write_tsv(as.data.frame(motif_ix@rowRanges), 
          file = paste0(path2project,
                        '/analysis/CHaRs_annotated.tsv.gz' ))


```


# Categorization of regions 

The following vulcano plot shows the peak scores for each region pre- and post-treatment
and in colors the categories to which we assigned each region.

```{r, splitting_genes_by_category}
peaks.df <- peaks.df %>%
  mutate(region_reg_type=ifelse(`log2FC: Group HCV_pre vs HCV_post`> 0,
                                'positive', 'negative')) %>%
  mutate(region_deg=(as.numeric(`q-value: Group HCV_pre vs HCV_post`)) < 0.05 )
is.na(peaks.df$region_deg) %>% sum
peaks.df$'category' <- sapply(1:nrow(peaks.df), 
                              function(i){
                                if ( is.na( peaks.df$region_deg[i] ) ) {
                                  return('NA')
                                } else {
                                  if ( peaks.df$region_deg[i] == FALSE ) {
                                  return('Scarred')
                                } 
                                 
                                if ( peaks.df$region_deg[i] == TRUE) {
                                  if ( peaks.df$region_reg_type[i] == 'positive') {
                                    return('Reversed')
                                    } 
                                  if ( peaks.df$region_reg_type[i] == 'negative') {
                                    return('Gained')
                                   } 
                                  }
                                }
                                
})

print('')
## Categorization of regions by the change in epigenetic landscape
peaks.df %>%
  ggplot(aes(x=`P1_HCV_pre`,
             y=`P1_HCV_post`,
             colour=category)) +
  geom_point() +
  geom_abline(linetype='dashed', color='black') +
  labs(x='HCV - DAA Pre-treatment Peak Score',
       y='HCV - DAA Post-treatment Peak Score',
       color='') +
  theme_classic()
  
```

```{r, regions_tf_targets_interactions}
interactions.df <- read_tsv(paste0(path2project,
                        '/analysis/CHaRs_annotated.tsv.gz' ))
#interactions.df <- as.data.frame(motif_ix@rowRanges)
interactions.df
```


```{r, adding_categories_to_regions}
## Checking order
any(interactions.df$start != peaks.df$Start & interactions.df$end != peaks.df$End)
interactions.df$'categories_postVsPreDAA' <- peaks.df$category
```


```{r, saving_categories, eval=FALSE}
write_tsv(
  peaks.df, 
  file = paste0(path2project, '/analysis/regions_categories_epigenetic_scar.tsv.gz')
)
```


# GRNBoost

We inferred a regulatory network using GRNBoost. The result of this
process is a list of interactions ranked by variable importance.

We ran the pySCENIC pipeline which contains GRNBoost
as the initial step, as show in the next chunk.


```{r, pySCENIC, eval=FALSE}
## Saving the matrix for Scenic
seurat <- readRDS(path2seurat)
mtx <- t(seurat@assays$SCT@scale.data)
colnames(mtx) <- gsub('--.*', '', colnames(mtx))
mtx <- mtx[, ! grepl('-|\\.', colnames(mtx))]
dim(mtx)
write.table(mtx, file = path2matrix, sep = '\t')

## The next commands should be run in the command line
## Please follow the instructions in the SCENIC pipeline here:
## https://github.com/hdsu-bioquant/pySCENIC_pipeline
## Modify the mtx variable in the config file to the path
## to where the matrix was stored, in this case, mtx = input/matrix.thime.tsv
## Then run
# > snakemake --configfile configs/config_nina_reprocessed.yml --cores 20

## Copy analysis to the project folder
dir.create(path = paste0(path2project, '/analysis'))
dir.create(path = paste0(path2project, '/analysis/scenic/'))
dir.create(path = paste0(path2project, '/analysis/scenic/daa_treatment'))

scenic.output <- list.files(paste0(path2scenic, '/output'), full.names = T)
scenic.output <- grep('sv', scenic.output, value = TRUE)
lapply(scenic.output,
       function(file_name){
         file.copy(from = file_name,
                   to =  paste0(path2project, '/analysis/scenic/daa_treatment'))
       })
list.files(paste0(path2project, '/analysis/scenic/daa_treatment'))

```

The following table shows the output of GRNBoost.

```{r, grn_boost}
grnboost <- read.table(paste0(path2project,
                              '/analysis/scenic/daa_treatment/adj.tsv'),
                       header = TRUE)
head(grnboost)
dim(grnboost)

  ## Number of ranked interactions per target
n_int <- 50

gboost.list <- grnboost %>% split.data.frame(f=grnboost$target)
gboost.list <- lapply(gboost.list, 
                      function(df) arrange(df, desc(importance)))
gboost.list <- lapply(gboost.list, 
                      function(df) head(df, n=n_int))
gboost.df <- do.call(rbind, gboost.list)
head(gboost.df)
dim(gboost.df)
```

# GRNBoost example

The next joystick plot shows the top 50 ranked TFs important to predict Tcf7.


```{r}
print('')
gboost.df %>%
  dplyr::filter(target=='TCF7') %>%
  mutate(rank=1:n()) %>%
  mutate(highlight=ifelse(rank<15, TRUE, FALSE)) %>%
  mutate(gene_label=ifelse(highlight==TRUE, TF, '')) %>%
  ggplot(aes(x=rank, y=importance, 
             label=gene_label,
             colour=highlight)) +
       geom_point() +
       geom_label_repel(max.overlaps = Inf) +
       scale_color_manual(values=c('black', 'red')) +
       theme_bw() +
       theme(legend.position = 'None')
       
```


# Getting TF-target interactions from regions

From the annotated regions using the previous methodology
we created a table of TF-targets interactions.

```{r, prunning_interactions}
## filtering regions that does not appears in the
## GRNBoost analysis
interactions.simp.df <- filter(interactions.df,
                               genes %in% gboost.df$target) 

## Extracting interactions
process_regions <- function(ints.df){
  target.tf.list <- lapply(
    1:nrow(ints.df), function(i) {
      tfs <- str_split(ints.df[i, 'TF'], 
                       pattern = ';')
      df <- data.frame(gene=ints.df$genes[i],
                       TF=tfs)
      colnames(df) <- c('gene', 'TF')
      df <- mutate(df, TF = toupper(TF))
      return(df)
    })
  TF_target.df <- do.call(rbind, target.tf.list)
  TF_target.df <- TF_target.df[! duplicated(TF_target.df),]
  return(TF_target.df)
}

TF_target.df.list <- list(
  gained=process_regions(
          filter(interactions.simp.df,
                 categories_postVsPreDAA %in% c('Gained'))),
  reversed=process_regions(
          filter(interactions.simp.df,
          categories_postVsPreDAA %in% c('Reversed'))),
  scarred=process_regions(
          filter(interactions.simp.df,
          categories_postVsPreDAA %in% c('Scarred')))
)
lapply(TF_target.df.list, head)
lapply(TF_target.df.list, dim)
dim(grnboost)
head(gboost.df)

## Intersecting GRN boost and regions
gboost.df <- dplyr::rename(gboost.df, gene=target)
granet.interactions.df.list <- lapply(
  TF_target.df.list, function(df){
    merged.df <- merge(df, gboost.df)
    merged.df <- dplyr::select(merged.df, gene, TF)
    merged.df <- merged.df[!duplicated(merged.df),]
    merged.df
})
#lapply(granet.interactions.df.list, function(df) filter(df, TF=='TCF7'))
#lapply(granet.interactions.df.list, dim)

## Getting list of interactions
interactions.list <- lapply(granet.interactions.df.list, 
                       function(df) {
                         df.processed <- mutate(df, interaction=paste0(TF, '--', gene))
                         interactions <- pull(df.processed, interaction)
                         interactions <- unique(interactions)
                         return(interactions)
})
lapply(interactions.list, length)

## Getting list of TFs
tfs.list <- lapply(granet.interactions.df.list, 
                       function(df) {
                         tfs <- pull(df, TF) %>% unique()
                         return(tfs)
})
lapply(tfs.list, length)


print('')
interactions_Venn <- ggVennDiagram(interactions.list) +
   theme(legend.position = 'none') +
   viridis::scale_fill_viridis() +
   labs(title='Number of interactions')

regulons_Venn <- ggVennDiagram(tfs.list) +
   theme(legend.position = 'none') +
   viridis::scale_fill_viridis() +
   labs(title = 'Regulons')

gridExtra::grid.arrange(
  interactions_Venn, regulons_Venn,
  ncol=2
)

```


```{r}
tfs.human <- readLines('https://raw.githubusercontent.com/aertslab/pySCENIC/master/resources/hs_hgnc_curated_tfs.txt')
interactions.TF_Only.list <- lapply(
  granet.interactions.df.list, 
  function(ints.df){
    df <- dplyr::filter(ints.df, TF %in% tfs.human)
    df <- dplyr::filter(df, gene %in% tfs.human)
    df <- unique(df)
    return(df)
})
lapply(interactions.TF_Only.list, dim)

dir.create(paste0(path2project, '/analysis/network'))
lapply(names(interactions.TF_Only.list),
       function(name) {
         write.table(interactions.TF_Only.list[name][[1]],
                     file = paste0(path2project, '/analysis/network/', name, '.tsv'),
                     sep = '\t', 
                     quote = FALSE, 
                     row.names = FALSE)
})

interactions.TF_Only.list
#saveRDS(interactions.TF_Only.list, 
#        file = paste0(path2project, 'analysis/interactions.TF_Only.list'))
```

```{r, eval=FALSE}
peaks.annotated <- readRDS(
  file = paste0(path2project,
                '/analysis/motifs_CHaRs_annotated.rds'))


tfs.human <- unique(tfs.human)
names(tfs.human) <- tfs.human

## Frequency of appearance of motifs in all categories
tfs.freq.overall <- sapply(tfs.human, function(tf) sum(grepl(tf, peaks.annotated$TF)))
## Frequency of appearance of motifs in gained regions
tfs.freq.gained <- sapply(tfs.human, 
                          function(tf) sum(grepl(tf, 
                                                 subset(peaks.annotated,
                                                      categories_postVsPreDAA=="Gained")$TF)))
## Frequency of appearance of motifs in reversed regions
tfs.freq.reversed <- sapply(tfs.human, 
                          function(tf) sum(grepl(tf, 
                                                 subset(peaks.annotated,
                                                        
                                                      categories_postVsPreDAA=="Reversed")$TF)))
## Frequency of appearance of motifs in scarred regions
tfs.freq.scarred <- sapply(tfs.human, 
                          function(tf) sum(grepl(tf, 
                                                 subset(peaks.annotated,
                                                      categories_postVsPreDAA=="Scarred")$TF)))


## Calculation of the fractions of motif appearance across categories
## Let P_cat = Frequency_cat / Total_cat
## We define the ratio Q_gained = P_gained / P_overall
P_overall <- tfs.freq.overall / nrow(peaks.annotated)
P_gained <- tfs.freq.gained / nrow(subset(peaks.annotated, 
                                          categories_postVsPreDAA=="Gained"))
P_reversed <- tfs.freq.reversed / nrow(subset(peaks.annotated,
                                            categories_postVsPreDAA=="Reversed"))
P_scarred <- tfs.freq.scarred / nrow(subset(peaks.annotated,
                                           categories_postVsPreDAA=="Scarred"))
Q_gained <- P_gained / P_overall
Q_reversed <- P_reversed / P_overall 
Q_Scarred <- P_scarred / P_overall

res.list <- list(Q_gained=Q_gained,
                 Q_reversed=Q_reversed,
                 Q_Scarred=Q_Scarred)
lapply(res.list, summary)

saveRDS(res.list, '../data/motif_ratio_by_category.rds', compress = TRUE)
```

```{r, fig.width=15, fig.height=6}
res.list <- readRDS('../data/motif_ratio_by_category.rds')
ratios_categories.plots.list <- lapply(names(res.list), 
       function(name){
         q_category <- res.list[name][[1]][!is.na(res.list[name][[1]])] %>% sort(decreasing = TRUE)
         tfs.df <- data.frame(tf=names(q_category),
                              q_category=q_category)
         tfs.df <- mutate(tfs.df, rank=1:n())
         tfs.df <- mutate(tfs.df, rescaled_scored=(q_category / max(q_category)))
         tfs.df <- mutate(tfs.df, gene_label=ifelse(rank<30, tf, ''))
         tfs.df %>%
           ggplot(aes(x=rank, y=log(q_category),
                      label=gene_label)) +
                geom_point() +
                theme_bw() +
                geom_label_repel(max.overlaps = Inf,
                     force = 40,
                     colour='Red') +
                xlim(-50, 600) +
                ylim(-0.3, 1) +
                labs(x='Rank', y='Scaled Quotient Motif by Category') +
                ggtitle(name)
})
gridExtra::grid.arrange(grobs=ratios_categories.plots.list, ncol=3)
```

## Heatmap Subpopulations Vs Gained/Reversed/Scarred


```{r, loading_maike_data}
seurat <- readRDS(paste0(path2project, 'data/maike2020/hofmann_hcv_seu.rds'))
seurat <- subset(seurat, nina_annotations %in% c('mem', 'exh', 'trans'))

seurat <- CreateSeuratObject(counts = seurat@assays$RNA@counts,
                             project = 'CD8 T cells', 
                             assay = 'RNA', 
                             min.cells = 1, 
                             min.features = 1, 
                             meta.data = seurat@meta.data)
seurat <- SCTransform(seurat)

seurat <- FindVariableFeatures(seurat)
seurat <- RunPCA(seurat, npcs = 20)
seurat <- RunUMAP(seurat, dims = 1:20)

```

```{r, heatmap_categories_x_cell_types}
peaks.annotated <- readRDS(
  file = paste0(path2project,
                '/analysis/motifs_CHaRs_annotated.rds'))
categories.df <- filter(peaks.annotated, genes != '')
genesByCategory <- split(categories.df$genes, categories.df$categories_postVsPreDAA)
genesByCategory <- genesByCategory[c('Gained', 'Reversed', 'Scarred')]
genesByCategory <- lapply(genesByCategory, unique)
genesByCategory <- lapply(genesByCategory,
                          function(x) intersect(x, rownames(seurat)))
genesByCategory$'Random genes' <- sample(rownames(seurat), size = 735)


mtxByCategory <- lapply(genesByCategory,
                        function(genes) FetchData(seurat,
                                                  vars = c('nina_annotations', 
                                                           genes)))
## Mean values
mtxByCategory <- lapply(mtxByCategory, function(mtx){
  df <- group_by(mtx, nina_annotations)
  summ.df <- summarise_all(df, mean)
  summ.df
})
## Plotting
boxplots.list <- lapply(names(mtxByCategory), function(category){
  mtxByCategory[category][[1]] %>%
    melt %>%
    ggplot(aes(x=nina_annotations, y=log(value))) +
    geom_jitter() +
    geom_boxplot() +
    ggtitle(category) +
    stat_compare_means(label.y = 1.5, label = "p.signif",
                       colour='red', size=5) +
    theme_bw() +
    theme(axis.text.x = element_text(hjust = 1, angle=45)) +
    labs(x='', y='Log(mean(gene expression))')
})
boxplot<- gridExtra::grid.arrange(grobs=boxplots.list, ncol=4)
plot(boxplot)


heatmap.mtx <- lapply(mtxByCategory,
       function(mtx){
         df <- dplyr::select(mtx, -nina_annotations)
         mean.vals <- apply(df, 1, mean)
         mean.vals
       }) %>%
       unlist() %>%
       matrix(ncol=3, byrow = TRUE)
rownames(heatmap.mtx) <- names(mtxByCategory)
colnames(heatmap.mtx) <- c('Mem', 'Trans', 'Exh')

heatmap.ggplot <- heatmap.mtx %>%
  t() %>%
  scale(center = TRUE) %>%
  t() %>%
  as.data.frame() %>%
  add_rownames(var = 'category') %>%
  melt() %>%
  ggplot(aes(x=variable,
             y=factor(category,
                      levels=c('Random genes',
                               'Scarred',
                               'Gained',
                               'Reversed')),
             fill=value)) +
         geom_tile(width=1,
                   height=1,
                   size=1,
                   colour='white') +
         scale_fill_gradient2(low = 'chartreuse4',
                              mid = 'Yellow',
                              high = 'Red') +
  coord_equal() +
  theme(panel.grid = element_blank(),
        panel.background = element_blank(),
        axis.ticks = element_blank()) +
  labs(x='', y='')
print('')
heatmap.ggplot
```

```{r, heatmap_genes_vs_cells}
head(categories.df)

## Filtering genes
categories.df <- filter(categories.df, categories_postVsPreDAA != 'NA')

## Trying only DEGs
Idents(seurat) <- seurat$nina_annotations
degs <- FindAllMarkers(seurat, 
                       logfc.threshold = 0.001, 
                       min.pct = 0.001, min.diff.pct = 0.001)
degs.significant <- filter(degs, p_val < 0.05)

dim(degs.significant)

## filtering genes by DEGs
genes.chars <- intersect(rownames(seurat), categories.df$genes) 
genes.chars <- genes.chars[!duplicated(genes.chars)]
genes.anns <- dplyr::select(categories.df, genes, categories_postVsPreDAA) %>%
                      filter(genes %in% genes.chars)
genes.split <- split(genes.anns$categories_postVsPreDAA, genes.anns$genes)
genes.unambigous.cat <- sapply(genes.split, function(x) length(unique(x)) < 2 )
genes.unambigous.cat <- names(genes.unambigous.cat)[genes.unambigous.cat]
genes.chars <- genes.chars[genes.chars %in% genes.unambigous.cat]
length(genes.chars)






## Removing genes with ambiguous categories


intersect(degs.significant, genes.chars) ## no intersection



## genes annotations
genes.anns <- genes.anns[!duplicated(genes.anns), ]
## Extract matrix from seurat 
mtx <- FetchData(seurat, 
                 vars = intersect(rownames(seurat), 
                                  genes.chars))
dim(mtx)

## cell annotations
mtx.ann <- seurat@meta.data
dim(mtx.ann)



row.anns <- rowAnnotation(cell_type=mtx.ann$nina_annotations)
#Heatmap(
#  mtx, show_row_names = FALSE, 
#  show_column_names = FALSE,
#  row_split = mtx.ann$nina_annotations,
#  column_split = genes.anns$categories_postVsPreDAA
#)
```


```{r, network_visualization, fig.width=15, fig.height=12}
interactions.TF_Only.list <- read_rds(file = paste0(path2project,
                                                    'analysis/interactions.TF_Only.list'))

interactions.TF_Only.list <- lapply(names(interactions.TF_Only.list), 
                                 function(name) mutate(interactions.TF_Only.list[name][[1]], 
                                                       category=name))
interactions.TF_Only.list <- lapply(interactions.TF_Only.list, 
                                    function(df) mutate(df, interaction=paste0(gene,'--', TF)))
interactions.tf.df <- do.call(rbind, interactions.TF_Only.list)
interactions.tf.df <- mutate(interactions.tf.df, 
                             color=plyr::mapvalues(interactions.tf.df$category,
                                                   from = c('gained', 'reversed', 'scarred'),
                                                   to = c('green', 'red', 'gray85')))
table(interactions.tf.df$category)

## interactions graph
edges.df <- mutate(interactions.tf.df, from=TF, to=gene)
vertices.df <- data.frame(genes=unique(c(interactions.tf.df$gene, interactions.tf.df$TF)))
vertices.df

## Importing to igraph
g <- graph_from_data_frame(d = edges.df, directed = TRUE, vertices = vertices.df)
V(g)$degree <- degree(g)

## Select vertices with highest degrees
v.degree <- degree(g)
v.sorted <- sort(v.degree, decreasing = TRUE) %>% head(30) %>% names()

g.subset <- induced_subgraph(g, v = v.sorted)
l <- layout_with_fr(g.subset)

dir.create(paste0(path2project, 'figures'))

pdf(paste0(path2project, '/figures/integrated_network.pdf'),
     width=15,
     height=12)
plot(g.subset, 
     layout=l,
     edge.arrow.size=0.1,
     vertex.size=(V(g.subset)$degree/max(V(g.subset)$degree)*27),
     vertex.label.cex=(V(g.subset)$degree/max(V(g.subset)$degree)*2),
     edge.width=0.5,
     vertex.frame.color='white')
dev.off()

plot(g.subset, 
     layout=l,
     edge.arrow.size=0.1,
     vertex.size=(V(g.subset)$degree/max(V(g.subset)$degree)*27),
     vertex.label.cex=(V(g.subset)$degree/max(V(g.subset)$degree)*2),
     edge.width=0.5,
     vertex.frame.color='white')
```

```{r, fig.width=12, fig.height=8}
names(interactions.TF_Only.list) <- c('Gained', 'Reversed', 'Scarred')
top.tfs.plots <- lapply(names(interactions.TF_Only.list), 
       function(name){
         df <- interactions.TF_Only.list[name][[1]]
         df %>% dplyr::select(TF, category) %>% table() %>%
           as.data.frame() %>%
           arrange(desc(Freq)) %>%
           head(20) %>%
           mutate(TF=factor(TF, rev(TF))) %>%
           ggplot(aes(x=TF, y=Freq)) +
                geom_bar(stat = 'identity',
                         fill='steelblue') +
                coord_flip() +
                labs(x='', y='Outgoing edges') +
                theme_bw() +
                ggtitle(name)
       })
gridExtra::grid.arrange(grobs=top.tfs.plots, ncol=3)
```


```{r, scores_heatmap, fig.width=4, fig.height=4.5}
res.list.filtered <- lapply(res.list, 
                            function(x) {
                              y <- x
                              y[is.na(y)] <- 0
                              return(y)
})

lapply(res.list, length)
q.df <- matrix(data = unlist(res.list.filtered), 
               ncol = length(res.list$Q_gained), 
               nrow = length(res.list),
               byrow = TRUE)
q.df[, 1:3]
colnames(q.df) <- names(res.list$Q_gained)
rownames(q.df) <- c('Gained', 'Reversed', 'Reversed')

scores.mtx <- t(q.df)
scores.mtx <- scores.mtx[apply(scores.mtx, 1, function(x) sum(x) > 0),]
scores.mtx <- log(scores.mtx)
Heatmap(scores.mtx, 
        show_row_names = FALSE, 
        row_dend_reorder = FALSE, 
        show_row_dend = FALSE,
        show_column_dend = FALSE,
        col = viridis(n=100, option = 'magma'))

```



```{r, eval=FALSE}
library(AUCell)
lapply(granet.interactions.df.list, head)

## Extracting all regulons
regulons <- lapply(granet.interactions.df.list, 
                   function(df) 
                     split(df$gene, f = df$TF)) %>% unlist(recursive = FALSE)

lapply(regulons, length)

## Saving regulons 
saveRDS(regulons, 
        '/media/ag-cherrmann/cramirez/tcd8ExscSeq/analysis/granet_regulons_categories_list.rds', 
        compress = TRUE

)

## Filtering regulons

#regulons <- regulons[sapply(regulons, function(reg) length(reg) > 5)]
  
## Getting cells with annotations
## Loading seurat object with integrated data Miller-Hoffmann
seurat.cell_types  <- readRDS(paste0(path2project,'data/maike2020/hofmann_hcv_seu.rds'))
seurat.cell_types <- subset(seurat.cell_types, nina_annotations %in% c('mem', 'trans', 'exh'))
## Visualization of cells
DimPlot(seurat.cell_types,
        group.by = 'nina_annotations')
mtx.cell_types <- seurat.cell_types@assays$integrated@scale.data

genes <- unique(rownames(mtx.cell_types))

#cells_rankings <- AUCell_buildRankings(mtx.cell_types)

#geneSets <- unlist(regulons, recursive = FALSE)
#cells_AUC <- AUCell_calcAUC(regulons, cells_rankings, 
#                            aucMaxRank=nrow(cells_rankings)*0.05)
#auc.mtx <- getAUC(cells_AUC)

#col_anns <- HeatmapAnnotation(cell_type = seurat.cell_types$nina_annotations)
#row_anns <- rowAnnotation(category = gsub('\\..*', '', rownames(auc.mtx)))

#Heatmap(auc.mtx, 
#        top_annotation = col_anns, 
#        left_annotation = row_anns,
#        column_split = seurat.cell_types$nina_annotations,
#        #row_split = gsub('\\..*', '', rownames(auc.mtx)),
#        show_row_names = FALSE,
#        show_column_names = FALSE)

#DimPlot(subset(seurat.cell_types, nina_annotations %in% c('mem', 'trans', 'exh')),
#        group.by = 'nina_annotations')

## Scoring cells
#seurat.cell_types$nina_annotations %>% unique()


## Selection of the TFs using a random forest predicting
## memory vs terminally phenotypes
#library(randomForest)

#auc.df <- t(auc.mtx) %>% as.data.frame()
#class(auc.df)
#auc.df <- mutate(auc.df, cell_type = seurat.cell_types$nina_annotations)
  
#n <- 4
#rforest <- randomForest(as.factor(cell_type)~., 
#                        data = auc.df) 
#varImpPlot(rforest)
#tfs <- rforest$importance %>% 
#               as.data.frame() %>%
#                arrange(desc(MeanDecreaseGini)) %>%
#                rownames()
#top_n_tfs <- tfs[1:4]
#top_n_tfs

#seurat.cell_types <- AddMetaData(seurat.cell_types, auc.df)

#FeaturePlot(subset(seurat.cell_types, nina_annotations %in% c('exh', 'mem')), 
#            features = 'scarred.PRDM1', pt.size = 2.5) +
#      scale_color_viridis()

#seurat.cell_types@meta.data %>%
#  ggplot(aes(x=nina_annotations,
#             y=log(`scarred.PRDM1`),
#             fill=nina_annotations)) +
#      geom_boxplot(draw_quantiles = 0.5, width = 0.5) +
#      theme_bw() +
#      labs(x='', y='Log(Signature AUC Score)', title = 'scarred PRDM1') +
#      theme(legend.position = 'none',
#            axis.text.x = element_text(angle=45, hjust=1)) +
#      stat_compare_means(aes(group = nina_annotations), 
#                         label = "p.signif", 
#                         label.y = 0.1,
#                         label.x = 2) 

#df.res <- seurat.cell_types@meta.data %>%
#  dplyr::select(nina_annotations, matches('gained|reversed|scarred')) %>%
#  group_by(nina_annotations) %>%
#  summarise_all(mean)

#df.res.t <- dplyr::select(df.res, -nina_annotations) %>% t
#colnames(df.res.t) <- df.res$nina_annotations
#df.res.t <- as.data.frame(df.res.t)
#df.res.t %>%
#  mutate(category=gsub('\\..*', '', rownames(df.res.t))) %>%
#  ggplot(aes(x=mem, y=exh, color=category)) +
#       geom_point() +
#       geom_abline(slope = 1)

auc.mtx <- readRDS('/media/ag-cherrmann/cramirez/tcd8ExscSeq/analysis/dta/auc_mtx.rds')
tfa <- CreateAssayObject(data = auc.mtx, min.cells = 0, min.features = 0)
seurat.cell_types[["regulon_"]] <- tfa
Idents(seurat.cell_types) <- seurat.cell_types$nina_annotations
diff.tfa <- FindMarkers(seurat.cell_types, 
               assay = 'regulon_', 
               group.by = 'nina_annotations', 
               logfc.threshold = 0, 
               min.pct = 0, 
               ident.1 = 'mem', ident.2 = 'exh')

#dir.create('/media/ag-cherrmann/cramirez/tcd8ExscSeq/analysis/dta')
#writexl::write_xlsx(
#        rownames_to_column(diff.tfa, var = 'regulon'),
#        '/media/ag-cherrmann/cramirez/tcd8ExscSeq/analysis/dta/differential_activity.xlsx'
#)


```



```{r, dta_vis, eval=FALSE}
print('')
pdf('/media/ag-cherrmann/cramirez/tcd8ExscSeq/figures/dta_vulcano_plot.pdf',
    height = 4, width = 6.5)
diff.tfa %>%
  mutate(category=gsub('\\..*', '', rownames(diff.tfa))) %>%
  mutate(tf=gsub('.*\\.', '', rownames(diff.tfa))) %>%
  arrange(desc(abs(avg_log2FC))) %>%
  mutate(rank=1:n()) %>%
  mutate(tf_label=ifelse(rank<20|tf=='BACH2', tf, '')) %>%
  ggplot(aes(x=avg_log2FC, 
             y=-log10(p_val),
             colour=category,
             label=tf_label)) +
      geom_point(size=3) +
      geom_point(shape=1, size=3, color='black') +
      geom_text_repel() +
      geom_hline(yintercept = -log10(0.05), 
                 linetype = 'dashed') +
      theme_bw() +
      labs(title = 'Differential Regulon Activity',
           subtitle = 'Mem Vs Exh',
           color='',
           x='Mean Log2(FC)',
           y='-Log10(p-Value)')
dev.off()


#FeaturePlot(subset(seurat.cell_types, nina_annotations %in% c('exh', 'mem')), 
#            features = 'reversed.EOMES', pt.size = 3) +
#         scale_color_viridis()
#DimPlot(subset(seurat.cell_types, nina_annotations %in% c('exh', 'mem')),
#        group.by = 'nina_annotations')
```

## DTA comparing cured Vs chronic

```{r, fig.width=11, fig.height=6}
seu <- readRDS(
        '/media/ag-cherrmann/cramirez/tcd8ExscSeq/data/maike2020/hofmann_hcv_seu.rds'
)
seu[[]]$orig.ident %>% unique()
seu[[]]$nina_annotations %>% unique()
table(seu[[]]$orig.ident, seu[[]]$nina_annotations)

auc.mtx <- readRDS('/media/ag-cherrmann/cramirez/tcd8ExscSeq/analysis/dta/auc_mtx.rds')
auc.mtx %>% dim; seu
seu$'condition' <- ifelse(is.na(seu$nina_annotations), 'cured', 'chronic') 
tfa <- CreateAssayObject(data = auc.mtx, min.cells = 0, min.features = 0)
tfa
seu[["regulon_"]] <- tfa
Idents(seu) <- seu$condition
diff.tfa.HUB0044 <- FindMarkers(subset(seu, orig.ident %in% c('HUBF0044BL', 
                                                      'UBF0044SORT3BFU')), 
               assay = 'regulon_', 
               group.by = 'condition', 
               logfc.threshold = 0, 
               min.pct = 0, 
               ident.1 = 'cured', ident.2 = 'chronic')



diff.tfa.HUB0044.plot <- diff.tfa.HUB0044 %>%
  mutate(category=gsub('\\..*', '', rownames(diff.tfa.HUB0044))) %>%
  mutate(tf=gsub('.*\\.', '', rownames(diff.tfa.HUB0044))) %>%
  arrange(desc(abs(avg_log2FC))) %>%
  mutate(rank=1:n()) %>%
  mutate(tf_label=ifelse(rank<20|tf=='BACH2', tf, '')) %>%
  ggplot(aes(x=avg_log2FC, 
             y=-log10(p_val),
             colour=category,
             label=tf_label)) +
      geom_point(size=3) +
      geom_point(shape=1, size=3, color='black') +
      geom_text_repel() +
      geom_hline(yintercept = -log10(0.05), 
                 linetype = 'dashed') +
      theme_bw() +
      labs(title = 'Differential Regulon Activity',
           subtitle = 'Cured Vs Chronic - HUBF0044',
           color='',
           x='Mean Log2(FC)',
           y='-Log10(p-Value)')




diff.tfa.HUBF0709 <- FindMarkers(subset(seu, orig.ident %in% c('HUBF0709BL', 
                                                      'HUBF0709FUDay2')), 
               assay = 'regulon_', 
               group.by = 'condition', 
               logfc.threshold = 0, 
               min.pct = 0, 
               ident.1 = 'cured', ident.2 = 'chronic')



diff.tfa.HUBF0709.plot <- diff.tfa.HUBF0709 %>%
  mutate(category=gsub('\\..*', '', rownames(diff.tfa.HUBF0709))) %>%
  mutate(tf=gsub('.*\\.', '', rownames(diff.tfa.HUBF0709))) %>%
  arrange(desc(abs(avg_log2FC))) %>%
  mutate(rank=1:n()) %>%
  mutate(tf_label=ifelse(rank<20|tf=='BACH2', tf, '')) %>%
  ggplot(aes(x=avg_log2FC, 
             y=-log10(p_val),
             colour=category,
             label=tf_label)) +
      geom_point(size=3) +
      geom_point(shape=1, size=3, color='black') +
      geom_text_repel() +
      geom_hline(yintercept = -log10(0.05), 
                 linetype = 'dashed') +
      theme_bw() +
      labs(title = 'Differential Regulon Activity',
           subtitle = 'Cured Vs Chronic - HUBF0709',
           color='',
           x='Mean Log2(FC)',
           y='-Log10(p-Value)')


pdf('/media/ag-cherrmann/cramirez/tcd8ExscSeq/figures/dta_vulcano_plot_conditions.pdf',
    width = 12, height = 5)
diff.tfa.HUB0044.plot + diff.tfa.HUBF0709.plot
dev.off()

DefaultAssay(seu) <- 'regulon_'
##seu <- NormalizeData(seu)


dplot.hub0044 <- DotPlot(subset(seu, orig.ident %in% c('HUBF0044BL', 
                                                      'UBF0044SORT3BFU')), 
        features = c('reversed.BATF3',
                     'reversed.BACH2',
                     'gained.ZSCAN29',
                     'gained.ATF3',
                     'gained.MECOM',
                     'scarred.BACH2'), 
        cols = 'RdYlGn', dot.scale = 10) +
        coord_flip() +
        labs(x='', y='', title = 'HUBF0044')


## HUBF0709 check
dplot.hub0709 <- DotPlot(subset(seu, orig.ident %in% c('HUBF0709BL', 'HUBF0709FUDay2')), 
        features = c('reversed.BATF3',
                     'reversed.TGIF1',
                     'reversed.BHLHE40',
                     'scarred.NR3C2',
                     'reversed.EOMES',
                     'reversed.BACH2'), 
        cols = 'RdYlGn', dot.scale = 10) +
        coord_flip() +
        labs(x='', y='', title = 'HUBF0709')

dplot.hub0044 + dplot.hub0709
```


```{r, dta_pooling_patients}
diff.tfa.condition <- FindMarkers(
        subset(seu, condition %in% c('chronic', 'cured')), 
               assay = 'regulon_', 
               group.by = 'condition', 
               logfc.threshold = 0, 
               min.pct = 0, 
               ident.1 = 'cured', ident.2 = 'chronic')



diff.tfa.condition.plot <- diff.tfa.condition %>%
  mutate(category=gsub('\\..*', '', rownames(diff.tfa.condition))) %>%
  mutate(tf=gsub('.*\\.', '', rownames(diff.tfa.condition))) %>%
  arrange(desc(abs(avg_log2FC))) %>%
  mutate(rank=1:n()) %>%
  mutate(tf_label=ifelse(rank<20|tf=='BACH2', tf, '')) %>%
  ggplot(aes(x=avg_log2FC, 
             y=-log10(p_val),
             colour=category,
             label=tf_label)) +
      geom_point(size=3) +
      geom_point(shape=1, size=3, color='black') +
      geom_text_repel() +
      geom_hline(yintercept = -log10(0.05), 
                 linetype = 'dashed') +
      theme_bw() +
      labs(title = 'Differential Regulon Activity',
           subtitle = 'Cured Vs Chronic',
           color='',
           x='Mean Log2(FC)',
           y='-Log10(p-Value)')


pdf('/media/ag-cherrmann/cramirez/tcd8ExscSeq/figures/dta_vulcano_plot_conditions_pooled_patients.pdf',
    width = 8, height = 5)
diff.tfa.condition.plot
dev.off()

dplot.conditions <- DotPlot(seu, 
        features = c('reversed.BATF3',
                     'reversed.TGIF1',
                     'scarred.MSX1',
                     'scarred.KLF6',
                     'gained.NR1D2',
                     'reversed.BACH2'), 
        cols = 'RdYlGn', dot.scale = 12) +
        coord_flip() +
        labs(x='', y='', title = 'HUBF0709')

print('')
dplot.conditions
```

# Comparing cured Vs memory

```{r, cured_Vs_memory}
seu$cell_type <- as.character(seu$nina_annotations)
seu$cell_type[is.na(seu$cell_type)] <- 'Cells from cured'
diff.tfa.curedVsMem <- FindMarkers(
        subset(seu, cell_type %in% c('Cells from cured', 'mem')), 
               assay = 'regulon_', 
               group.by = 'cell_type', 
               logfc.threshold = 0, 
               min.pct = 0, 
               ident.1 = 'Cells from cured', 
               ident.2 = 'mem')

diff.tfa.curedVsMem.plot <- diff.tfa.curedVsMem %>%
  mutate(category=gsub('\\..*', '', rownames(diff.tfa.curedVsMem))) %>%
  mutate(tf=gsub('.*\\.', '', rownames(diff.tfa.curedVsMem))) %>%
  arrange(desc(abs(avg_log2FC))) %>%
  mutate(rank=1:n()) %>%
  mutate(tf_label=ifelse(rank<20|tf=='BACH2', tf, '')) %>%
  ggplot(aes(x=avg_log2FC, 
             y=-log10(p_val),
             colour=category,
             label=tf_label)) +
      geom_point(size=3) +
      geom_point(shape=1, size=3, color='black') +
      geom_text_repel() +
      geom_hline(yintercept = -log10(0.05), 
                 linetype = 'dashed') +
      theme_bw() +
      labs(title = 'Differential Regulon Activity',
           subtitle = 'Cured Vs Memory',
           color='',
           x='Mean Log2(FC)',
           y='-Log10(p-Value)')


diff.tfa.curedVsMem.plot

dplot.curedVsMem <- DotPlot(
        subset(seu, cell_type %in% c('Cells from cured', 'mem')), 
        features = c('reversed.EBF1',
                     'reversed.AR',
                     'reversed.FOXJ2',
                     'gained.BACH2',
                     'reversed.NR2E3',
                     'scarred.NR3C2',
                     'reversed.TGIF1',
                     'reversed.BATF3'), 
        cols = 'RdYlGn', dot.scale = 12,
        group.by = 'cell_type') +
        coord_flip() +
        labs(x='', y='')

dplot.curedVsMem
```


```{r, cured_Vs_exh}
seu$cell_type <- as.character(seu$nina_annotations)
seu$cell_type[is.na(seu$cell_type)] <- 'Cells from cured'
diff.tfa.curedVsExh <- FindMarkers(
        subset(seu, cell_type %in% c('Cells from cured', 'exh')), 
               assay = 'regulon_', 
               group.by = 'cell_type', 
               logfc.threshold = 0, 
               min.pct = 0, 
               ident.1 = 'Cells from cured', 
               ident.2 = 'exh')

diff.tfa.curedVsExh.plot <- diff.tfa.curedVsExh %>%
  mutate(category=gsub('\\..*', '', rownames(diff.tfa.curedVsExh))) %>%
  mutate(tf=gsub('.*\\.', '', rownames(diff.tfa.curedVsExh))) %>%
  arrange(desc(abs(avg_log2FC))) %>%
  mutate(rank=1:n()) %>%
  mutate(tf_label=ifelse(rank<20|tf=='BACH2', tf, '')) %>%
  ggplot(aes(x=avg_log2FC, 
             y=-log10(p_val),
             colour=category,
             label=tf_label)) +
      geom_point(size=3) +
      geom_point(shape=1, size=3, color='black') +
      geom_text_repel() +
      geom_hline(yintercept = -log10(0.05), 
                 linetype = 'dashed') +
      theme_bw() +
      labs(title = 'Differential Regulon Activity',
           subtitle = 'Cured Vs Exhausted',
           color='',
           x='Mean Log2(FC)',
           y='-Log10(p-Value)')

seu$cell_type %>% table()
diff.tfa.curedVsExh.plot

dplot.curedVsExh <- DotPlot(
        subset(seu, cell_type %in% c('Cells from cured', 'exh')), 
        features = c('scarred.EOMES',
                     'scarred.NFIC',
                     'gained.CREM',
                     'scarred.BACH2',
                     'gained.BACH2',
                     'reversed.FOXJ2',
                     'reversed.AR',
                     'reversed.EBF1'), 
        cols = 'RdYlGn', dot.scale = 12,
        group.by = 'cell_type') +
        coord_flip() +
        labs(x='', y='')

dplot.curedVsExh

pdf('/media/ag-cherrmann/cramirez/tcd8ExscSeq/figures/dotplot_check_dta_curedVsMemOrExh.pdf',
    width = 12, height = 4)
dplot.curedVsMem + dplot.curedVsExh
dev.off()
```


```{r, plotting_dta_plots, fig.width=15, fig.height=5}
pdf('/media/ag-cherrmann/cramirez/tcd8ExscSeq/figures/dta_vulcano_CuredVsMemOrExh.pdf',
    width = 15, height = 5)
diff.tfa.curedVsMem.plot + diff.tfa.curedVsExh.plot
dev.off()

diff.tfa.curedVsMem.plot + diff.tfa.curedVsExh.plot
```


## Variability in cured vs chronic samples


```{r, fig.width=8}
path2seurat <- '/media/ag-cherrmann/cramirez/tcd8ExscSeq/data/Thimme_reprocessed/seurat.thimme.rds'
seurat <- readRDS(path2seurat)
seurat$condition %>% unique()
DefaultAssay(seurat) <- 'RNA'
pca <- FetchData(seurat, vars = c('condition', 
                                  colnames(seurat@reductions$pca)), slot='count')

pca[,1:30] %>%
  group_by(condition) %>%
  summarise_all(var) %>%
  melt() %>%
  ggplot(aes(x=variable, y=value, colour=condition)) +
      geom_point() +
      theme_bw() +
      theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
      labs(x='PC', y='Variance')
```



```{r, node_importance}

# Extracting the vertices
GraphVertices <- V(My_graph)        

# Calculation of Hubness score
Hubness.score <- hubness.score(graph = g,     
                                   vertices = GraphVertices, 
                                   directed = FALSE, mode = "all",
                                   loops = TRUE, scaled = TRUE)


# Visualizing the graph based on IVI values
graph.vis <- cent_network.vis(graph = g,
                                     cent.metric = Hubness.score,
                                     directed = TRUE,
                                     plot.title = "TF Interactions Network",
                                     legend.title = "Hubness score", label.color = 'red')

print('')
graph.vis
```


```{r}
bach2.regs <- lapply(regulons, function(x) x$'BACH2')
bach2.regs.df.list <- lapply(names(bach2.regs), 
       function(name){
         df <- data.frame(TF='BACH2', category=name, target=bach2.regs[name][[1]])
         df }) 
bach2.regs.df <- do.call(rbind, bach2.regs.df.list) 
bach2.regs.df

write_csv(bach2.regs.df, file = paste0(path2project, '/analysis/bach2_rewiring.csv'))
```


```{r}
## Trying only DEGs
Idents(seurat) <- seurat$nina_annotations
degs <- FindAllMarkers(seurat, 
                       logfc.threshold = 0.001, 
                       min.pct = 0.001, min.diff.pct = 0.001)
degs.significant <- filter(degs, p_val < 0.05)

dim(degs.significant)




genes.anns <- dplyr::select(categories.df, genes, categories_postVsPreDAA) %>%
                      filter(genes %in% genes.chars)


## Removing genes with ambiguous categories
genes.split <- split(genes.anns$categories_postVsPreDAA, genes.anns$genes)
genes.unambigous.cat <- sapply(genes.split, function(x) length(unique(x)) < 2 )
genes.unambigous.cat <- names(genes.unambigous.cat)[genes.unambigous.cat]
genes.chars <- genes.chars[genes.chars %in% genes.unambigous.cat]
length(genes.chars)

## filtering genes by DEGs
genes.chars <- intersect(rownames(seurat), categories.df$genes) 
genes.chars <- genes.chars[!duplicated(genes.chars)]
intersect(degs.significant, genes.chars) ## no intersection


## genes annotations
genes.anns <- genes.anns[!duplicated(genes.anns), ]
## Extract matrix from seurat 
mtx <- FetchData(seurat, 
                 vars = intersect(rownames(seurat), 
                                  genes.chars))
dim(mtx)

## cell annotations
mtx.ann <- seurat@meta.data
dim(mtx.ann)



row.anns <- rowAnnotation(cell_type=mtx.ann$nina_annotations)
Heatmap(
  mtx, show_row_names = FALSE, 
  show_column_names = FALSE,
  row_split = mtx.ann$nina_annotations,
  column_split = genes.anns$categories_postVsPreDAA
)
```