---
title: "Regulatory network inference"
author: "Health Data Unit"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: hide
  pdf_document: default
---


# Intro

This report contain documentation of the analysis of CD8 T cells from chronic HCV
patients.


```{r}
library(rio)
library(ggplot2)
library(tidyverse)
library(motifmatchr)
library(TFBSTools)
library(JASPAR2020)
library(GenomicRanges)
library(BSgenome)
library(ggVennDiagram)
library(grid)
library(Gviz)
library(Homo.sapiens)
library(rGREAT)
library(UpSetR)

## Rmarkdown Settings
knitr::opts_chunk$set(
        warning = FALSE, 
        message = FALSE,
        cache = FALSE,
        fig.align = 'center'
)

set.seed(333)

path2project <- '/media/ag-cherrmann/cramirez/tcd8ExscSeq/'
```


# Reading peaks

* The data was taken from [Kathleen B. Y. et al, 2021](https://www.nature.com/articles/s41590-021-00979-1?proof=t#additional-information). 


The next table shows the universe of ChARs identified in all conditions.


```{r, loading_peaks}
peaks_url <- 'https://static-content.springer.com/esm/art%3A10.1038%2Fs41590-021-00979-1/MediaObjects/41590_2021_979_MOESM3_ESM.xlsx'

download.file(peaks_url, '/media/ag-cherrmann/cramirez/tcd8ExscSeq/data/41590_2021_979_MOESM3_ESM.xlsx')
peaks.df <- readxl::read_xlsx('/media/ag-cherrmann/cramirez/tcd8ExscSeq/data/41590_2021_979_MOESM3_ESM.xlsx', 
                              sheet = 'Supplementary Table 2')
dim(peaks.df)
head(peaks.df)
```

# The epigenetic scar

```{r, size_of_epigenetic_scar}
universe <- pull(peaks.df, ID)
preVspost.HCV <- filter(peaks.df, 
                        as.numeric(`q-value: Group HCV_pre vs HCV_post`) < 0.05) %>% pull(ID)
sets <- list(universe=universe,
             `Pre- Vs Post-Treatment`=preVspost.HCV)
```


The next Venn Diagram shows the number of CHaRs in the overall universe across all conditions 
(`r length(universe)`) and the number of differential peaks in the epigenetic scar which 
corresponds to the comparison of Pre- vs Post-DAA treatment (`r length(preVspost.HCV)`).


```{r}
ggVennDiagram(sets)
```

The next vulcano plot shows the epigenetic scar.

```{r, fig.height=4, fig.width=4}
## Visualization of the epigenetic scar
epi.scar.vulcano.plot <- peaks.df %>%
  ggplot(aes(x=`log2FC: Group HCV_pre vs HCV_post`,
             y=-log10(as.numeric(`q-value: Group HCV_pre vs HCV_post`)))) +
       geom_point(alpha=0.3, size=3) +
       geom_hline(yintercept = -log10(0.05), linetype='dashed', 
                  color='red') +
       theme_classic() +
       labs(x='Log2FC: HCV pre Vs post', y='-Log10(q-Value)') +
       annotation_custom(
         grobTree(textGrob('q-Value=0.05', x=0.2, y=0.1),
                gp=gpar(col='red'))
       )
       
       
epi.scar.vulcano.plot
```

# Scanning motifs in CHaRs



```{r}
## Constructing the set of PFmatrix containing all the motifs in jaspar2020
opts <- list()
opts[["all_versions"]] <- TRUE
PFMatrixList <- getMatrixSet(JASPAR2020, opts)

## Converting regions to GRanges
peaks <- peaks.df
#peaks <- filter(peaks.df, 
#                as.numeric(`q-value: Group HCV_pre vs HCV_post`) < 0.05)
ranges <- makeGRangesFromDataFrame(peaks, 
                                   ignore.strand = TRUE, 
                                   seqnames.field = 'Chr', 
                                   start.field = 'Start',
                                   end.field = 'End')


motif_ix <- matchMotifs(PFMatrixList, ranges, genome = "hg19", out = 'scores') 

saveRDS(motif_ix, paste0(path2project,
                         '/analysis/motifs_CHaRs_preVsPost_treatment.rds'),
        compress = TRUE)
saveRDS(ranges, paste0(path2project, 
                       'analysis/ranges_CHaRs_preVsPost_treatment.rds'),
        compress = TRUE)
dim(motifScores(motif_ix))
head(motifScores(motif_ix))
```

# Plotting an example of ChAR 

The next track plot show an interval of 100 kb around the IFNG gene.

```{r}
ranges <- readRDS(file = paste0(path2project,
                         '/analysis/ranges_CHaRs_preVsPost_treatment.rds'))
motifs_ix <- readRDS(file = paste0(path2project,
                         '/analysis/motifs_CHaRs_preVsPost_treatment.rds'))
from <- 68548548 - 100000 
to <- 68553520 + 100000
ranges.12 <- ranges[seqnames(ranges)=="chr12" &
                      from < start(ranges)  &
                         end(ranges) < to]
atrack <- AnnotationTrack(ranges.12)
gtrack <- GenomeAxisTrack()
chr <- 'chr12'
gen <- 'hg19'
itrack <- IdeogramTrack(genome = gen, chromosome = chr)


knownGenes <- UcscTrack(genome = "hg19", chromosome = "chr12", 
                        track = "knownGene", from = from, to = to,
                        trackType = "GeneRegionTrack", 
                        rstarts = "exonStarts", rends = "exonEnds", 
                        gene = "name", symbol = "name", 
                        transcript = "name", strand = "strand", 
                        fill = "#8282d2", name = "UCSC Genes")

plotTracks(list(itrack, atrack, gtrack, knownGenes))
```

# Annotating regions with genes around regions



```{r, manual_annotation}
## Construction of the annotated reference
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene::TxDb.Hsapiens.UCSC.hg19.knownGene
broads <- GenomicFeatures::genes(txdb)
subject <- broads[ seqnames(broads) %in% seqlevels(ranges) ]

## Construction of the gene annotations
TxDb(Homo.sapiens) <- txdb
tx.list <- transcriptsBy(Homo.sapiens, columns = "SYMBOL")
tx <- unlist(tx.list)

## Function to annotate regions to genes around 1, 3, 10, 50 and 100 kbs
getNeighborsGenes <- function(grange, txDb_reference, dist){
  tx.chr <- txDb_reference[seqlevels(txDb_reference) %in% seqlevels(grange)]
  distances <- GenomicRanges::distance(grange, tx.chr)
  names(distances) <- unlist(tx.chr$SYMBOL)
  distances <- distances[!is.na(distances) & ! is.na(names(distances))]
  return(distances[abs(distances)<dist])
}

distances <- c(1000, 3000, 5000, 10000, 50000)
neighbors.list <- lapply(distances, 
       function(distance_val){
         lapply(1:length(ranges), 
                    function(i){ 
                      getNeighborsGenes(ranges[i], 
                                        txDb_reference = tx, 
                                        dist = distance_val)})
})
names(neighbors.list) <- paste0('neighborGenes_', distances)
#lapply(neighbors.list, head)

granges.genes <- lapply(neighbors.list, 
                        function(res){
                          genes <- res %>% unlist() %>% names() %>% unique()
                        })
names(granges.genes) <- paste0('genes_around_', c(1, 3, 5, 10, 50), '_kb')

saveRDS(granges.genes)
```


```{r}
upset(fromList(granges.genes), nsets = 20)
```


```{r, distances}
distances.filtered <- neighbors.list$neighborGenes_1000 %>% 
                             unlist() 
sum(distances.filtered==0)
hist(distances.filtered[distances.filtered> 0])
```


```{r, annotation_using_great, fig.height=5, fig.width=14, eval=FALSE}
distances.rgreat <- c(1, 3, 5, 10, 50)
names(distances.rgreat) <- paste0('rgreat_', distances.rgreat, '_kb')
## Testing different windows
job.list <- lapply(distances.rgreat, 
                   function(distance){
                     job <- submitGreatJob(ranges,
                      species = 'hg19', 
                      rule = "twoClosest", 
                      adv_twoDistance = distance)
                     return(job)
                   })
job.list <- lapply(job.list, function(job){
              great_job <- job
              res <- plotRegionGeneAssociationGraphs(great_job, plot=FALSE)
              return(great_job)
})

#names(job.list)
#job.list$rgreat_1_kb@association_tables$all$gene %>% unique() %>% length()   
#job.list$rgreat_10_kb@association_tables$all[
#  ! is.na(job.list$rgreat_1_kb@association_tables$all$distTSS) &
#  job.list$rgreat_1_kb@association_tables$all$distTSS < 1000,
#] %>% pull(gene) %>% unique() %>% length()
plotRegionGeneAssociationGraphs(job.list$rgreat_1_kb)
```

```{r, eval=FALSE}
rgreat.genes <- lapply(1:length(job.list), 
                       function(index){
                         regions.tb <- job.list[index][[1]]@association_tables$all
                         regions.filtered <- regions.tb[ ! is.na(regions.tb$distTSS) & 
                                                           regions.tb$distTSS < distances.rgreat[index], ] 
                         genes <- regions.filtered %>% pull(gene) %>% unique()
                         return(genes)
                       })

upset(fromList(rgreat.genes), 
      nsets = 20, sets = names(rgreat.genes), 
      keep.order = TRUE)
```

```{r, comparing_results, eval=FALSE}
lapply(neighbors.list, head)


neighbors.genes <- c(granges.genes, rgreat.genes)

names(neighbors.genes) <- c(paste0('gRanges_', c(1, 3, 5, 10, 50), '_kb'),
                            paste0('rGreat_', c(1, 3, 5, 10, 50), '_kb'))

names(rgreat.genes) <- names(granges.genes)

upset(fromList(neighbors.genes), 
      nsets = 20, sets = names(neighbors.genes), keep.order = TRUE)
```



```{r, match_motif_annotation}
motif_ix 
mapped_genes <- ''

lapply(neighbors.list$neighborGenes_1000,
        function(genes){
          names(genes)
})
```
